File: cpp/1_twoSum.cpp
#include <map>
using namespace std;
// 1.brute force需要O(n2)的遍历
// 2.思路是一次遍历储存下已经遍历的值，在这些值里查找就可以避免重复遍历

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int,int> Map;
        vector<int> res;
        typedef pair<int,int> p;

        auto item_i = nums.at(0);
        Map.insert(p(item_i,0));

        for (auto j=1; j<nums.size();j++)
        {
            if (nums[j]==target-item_i)
            {
                res={0,j};
                break;
            }
            else
            {
                if (Map.count(target-nums[j])>0)
                {
                    res = {j,Map.at(target-nums[j])};
                    break;
                }
                Map.insert(p(nums[j],j));
            }
        }
        return res;
    }
};

File: java/1_twoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        var map = new HashMap<Integer,Integer>(); // number-location pairs
        var item_i = nums[0];
        var res = new int[2];
        map.put(item_i,0);

        for (var i=1;i<nums.length;i++)
        {
            if (item_i==target-nums[i])
            {
                res[0]=i;
                res[1]=map.get(item_i);
                break;
            }
            else
            {
                if (map.containsKey(target-nums[i]))
                {
                    res[0] = map.get(target-nums[i]);
                    res[1] = i;
                    break;
                }
                map.put(nums[i],i);
            }
        } 
        return res;
    }
}

